File: web/lib/shipment-status.ts
import { t, type Lang } from "@/lib/i18n";

/**
 * Every status Biteship documents, in happy-path order first and then the
 * branches that mean something went wrong.
 */
export const SHIPMENT_STATUSES = [
  "confirmed",
  "scheduled",
  "allocated",
  "picking_up",
  "picked",
  "in_transit",
  "dropping_off",
  "delivered",
  "on_hold",
  "rejected",
  "courier_not_found",
  "cancelled",
  "return_in_transit",
  "returned",
  "disposed",
] as const;

export type ShipmentStatus = (typeof SHIPMENT_STATUSES)[number];

/**
 * Statuses that mean the parcel is not going to arrive without someone
 * intervening. `on_hold` is deliberately absent: it resolves on its own and
 * flagging it would train admins to ignore the badge.
 */
export const SHIPMENT_FAILURE_STATUSES: readonly ShipmentStatus[] = [
  "rejected",
  "courier_not_found",
  "cancelled",
  "return_in_transit",
  "returned",
  "disposed",
];

/**
 * Biteship's docs render these camelCase (`courierNotFound`) while payloads
 * use snake_case (`courier_not_found`), and most have never been seen on the
 * wire here. Folding both onto one key costs nothing and avoids a status
 * silently falling through to its raw code.
 */
export function normaliseShipmentStatus(raw: string): string {
  return raw
    .trim()
    .replace(/([a-z0-9])([A-Z])/g, "$1_$2")
    .toLowerCase();
}

export function shipmentStatusLabel(raw: string, lang: Lang): string {
  const key = normaliseShipmentStatus(raw);
  if (!(SHIPMENT_STATUSES as readonly string[]).includes(key)) return raw;
  // The dictionary is keyed by literal strings, so a computed key needs the
  // cast; the includes() guard above is what makes it safe.
  return t(lang, `shipment_status_${key}` as Parameters<typeof t>[1]);
}

export function isShipmentFailure(raw: string | null | undefined): boolean {
  if (!raw) return false;
  return (SHIPMENT_FAILURE_STATUSES as readonly string[]).includes(
    normaliseShipmentStatus(raw),
  );
}

File: backend/internal/service/shipment_status.go
package service

import "strings"

// ShipmentFailureStatuses are the Biteship statuses that mean the parcel is
// not going to arrive without someone intervening.
//
// on_hold is deliberately absent: it resolves on its own, and surfacing it as
// a failure would train admins to ignore the signal.
//
// This matters because the webhook never walks orders.status back (FR-C-15) —
// an order whose courier vanished still reads "shipped" everywhere else, so
// this list is the only thing that can surface it.
var ShipmentFailureStatuses = []string{
	"rejected",
	"courier_not_found",
	"cancelled",
	"return_in_transit",
	"returned",
	"disposed",
}

// shipmentStatusSpellings expands a canonical snake_case status into every
// spelling that could be sitting in the column.
//
// We store whatever Biteship sends, verbatim and unnormalised. Their docs
// render these statuses camelCase while payloads appear to use snake_case, and
// most have never been observed here at all — so a filter that matched only
// one spelling would quietly return nothing for the other. Matching both is
// cheap; guessing is not.
func shipmentStatusSpellings(canonical string) []string {
	parts := strings.Split(canonical, "_")
	if len(parts) == 1 {
		return []string{canonical}
	}
	camel := parts[0]
	for _, p := range parts[1:] {
		if p == "" {
			continue
		}
		camel += strings.ToUpper(p[:1]) + p[1:]
	}
	return []string{canonical, camel}
}

// ShipmentFailureStatusValues is what the repository filter is given: every
// failure status in every spelling it might have been stored under.
func ShipmentFailureStatusValues() []string {
	var out []string
	for _, s := range ShipmentFailureStatuses {
		out = append(out, shipmentStatusSpellings(s)...)
	}
	return out
}

// isShipmentFailureStatus matches one stored shipment_status against the
// failure set, in either spelling. Built from ShipmentFailureStatusValues so a
// status added to the list is matched here without a second edit — the same
// reason the repository filter reads from it rather than its own literal.
// A nil or empty status is not a failure: it means nothing has been reported.
func isShipmentFailureStatus(status *string) bool {
	if status == nil || *status == "" {
		return false
	}
	for _, s := range ShipmentFailureStatusValues() {
		if s == *status {
			return true
		}
	}
	return false
}
